semeio/workflows/ahm_analysis/ahmanalysis.py

```diff
@@ -0,0 +1,456 @@
+        if combination_length == 1:
+            # We have hit a case with two observation groups, now we only
+            # need individual group and all obs.
+            continue
+        for subset in itertools.combinations(key_map.keys(), combination_length):
+            obs_group = list(
+                itertools.chain.from_iterable([key_map[key] for key in subset])
+            )
+            if combination_length == len(key_map.keys()) - 1:
+                missing_obs = [x for x in key_map.keys() if x not in set(subset)]
+                assert len(missing_obs) == 1
+                name = f"All_obs-{missing_obs[0]}"
+            else:
+                name = "All_obs"
+            combinations[name.replace(":", "_")] = obs_group
+    return combinations
+
+
+def list_active_observations(df_update_log):
+
+    """To get the inactive vs total observation info."""
+    df_active = df_update_log[(df_update_log["status"] == "Active")]
+    return str(len(df_active)) + " active/" + str(len(df_update_log.index))
+
+
+def create_path(runpathdf, target_name):
+    """to create path for output data"""
+    return tempfile.mkdtemp()
+
+
+def list_observations_misfit(obs_keys, df_update_log, misfit_df):
+    """To get the misfit for total observations (active/inactive)."""
+    misfit_values = []
+    total_obs_nr = len(df_update_log[df_update_log.status.isin(["Active", "Inactive"])])
+    for key in obs_keys:
+        misfit_name = "MISFIT:" + key
+        misfit_values.append(misfit_df[misfit_name].mean())
+    return np.sum(misfit_values) / total_obs_nr
+
+
+def get_field_grid_char(input_grid):
+    """Get field grid characteristics/coordinates"""
+    input_grid = Path(input_grid).with_suffix("")
+    try:
+        mygrid = Grid(input_grid, fformat="eclipserun")
+        dataframeg = mygrid.dataframe(activeonly=False)
+        return dataframeg
+    except OSError as err:
+        raise OSError("A grid with .EGRID format is expected.") from err
+
+
+def calc_delta_grid(
+    all_input_post, all_input_prior, mygrid_ok, caseobs, mygrid_ok_short
+):
+    """calculate mean delta of field grid data"""
+    delta_post_prior = np.subtract(all_input_prior, all_input_post)
+    delta_post_prior = np.absolute(delta_post_prior)
+    # Can we avoid changing in place here?
+    mygrid_ok["Mean_delta"] = np.mean(delta_post_prior, axis=0)
+    df_mean_delta = mygrid_ok.groupby(["IX", "JY"])[["Mean_delta"]].mean().reset_index()
+    mygrid_ok_short["Mean_D_" + caseobs] = df_mean_delta["Mean_delta"]
+    return mygrid_ok_short
+
+
+def _get_parameter_keys(genkw_parameters, prior_keys):
+    """excluding duplicate transformed parameter keys"""
+    transformed_param_keys = []
+    for parameter in genkw_parameters:
+        for key in prior_keys:
+            if "_" + parameter + ":" in key:
+                transformed_param_keys.append(key)
+    return set(prior_keys).difference(transformed_param_keys)
```

Comment:
This function doesn't really get anything, it rather filters. It also contains a lot of context on how ert formats data. If we agree that what ert would do is to prefix a parameter with the transformation name, then this is a suggestion to make it more readable:

```
def _filter_on_prefix(list_of_strings, prefixes):
    """Returns the set of strings that has a match for any of the given prefixes"""
    return {
        string
        for string in list_of_strings
        if any(string.startswith(prefix) for prefix in prefixes)
    }
```

And then where it is used we can call it by:

`parameter_keys = _filter_on_prefix(list_of_strings=prior_data.keys(), prefixes=parameters)`

Now at least we don't have to wonder what `get_parameter_keys` does..

Its a suggestion at least

Reply:
suggestion implemented